marlowe/src/Language/Marlowe/Client.hs

```diff
@@ -105,9 +112,97 @@ data PartyAction
 
 type RoleOwners = AssocMap.Map Val.TokenName PubKeyHash
 
-marlowePlutusContract :: Contract () MarloweSchema MarloweError ()
+type MarloweContractState = [MarloweData]
+
+data ContractHistory
+        = None
+        | Created MarloweParams MarloweData
+        | Transition TransactionInput
+        | History MarloweParams MarloweData [TransactionInput]
+    deriving (Show,Generic)
+    deriving anyclass (FromJSON, ToJSON)
+
+instance Semigroup ContractHistory where
+    any <> None                                  = any
+    _ <> Created params md                       = History params md []
+    History params md inputs <> Transition input = History params md (inputs <> [input])
+    cur <> _                                     = cur
+
+instance Monoid ContractHistory where
+    mempty = None
+
+marloweFollowContract :: Contract ContractHistory MarloweFollowSchema MarloweError ()
+marloweFollowContract = do
+    params <- endpoint @"follow"
+    slot <- currentSlot
+    logDebug @String "Getting contract history"
+    history slot params
+    logDebug @String "Start following"
+    follow params (slot + 5)
+  where
+    follow params timeout = do
+        wr <- SM.waitForUpdateUntil (mkMarloweClient params) timeout
+        logDebug @String ("State " <> show (encode wr))
+        case wr of
+            ContractEnded -> do
+                logDebug @String "Contract Ended"
+                pure ()
+            Timeout _ -> do
+                logDebug @String "Contract Timeout"
+                follow params (timeout + 5)
+            WaitingResult marloweData -> do
+                logDebug @String (show marloweData)
+                follow params (timeout + 5)
```

Comment:
I am missing a call to `tell`  to update the state with the new history. I saw one in `updateHistoryFromTx`, but I understand that is called only once

Reply:
Good point.